File: stacks/ltcd_85.py
'''
LeetCode - problem 85
'''

from collections import deque 
class Solution:
  def left(self,arr,n):
    result = [-1]*n
    stack = deque()
    stack.append(0)
    for i in range(1,n):
      currH = arr[i] 
      if stack and arr[stack[-1]]<currH:
        result[i] = stack[-1] 
      else:
        while stack and arr[stack[-1]]>=currH:
          stack.pop() 
        if stack:
          result[i] = stack[-1] 
      stack.append(i)
    return result
    
  def right(self,arr,n):
    stack = deque()
    stack.append(n-1)
    result=[n]*n 
    for i in range(n-2,-1,-1):
      currH = arr[i]
      if stack and arr[stack[-1]]<currH:
        result[i] = stack[-1] 
      else:
        while stack and arr[stack[-1]]>=currH:
          stack.pop() 
        
        if stack:
          result[i] = stack[-1]
      stack.append(i)
    return result 
    
  def getMaxArea(self,arr,n):
    left = self.left(arr,n)
    right = self.right(arr,n)
    maxArea = 0 
    for i in range(n):
      currArea = ((right[i]-left[i])-1)*(arr[i])
      maxArea = max(maxArea,currArea)
    return maxArea
  
  def maximalRectangle(self,matrix):
    n = len(matrix)
    m = len(matrix[0])
    maxArea = 0 
    height=[0]*m 
    for i in range(n):
    
      for j in range(m):
        if matrix[i][j] == "0":
          height[j] = 0 
        else:
          height[j]+= int(matrix[i][j])
      maxArea = max(maxArea, self.getMaxArea(height,m))
    return maxArea

File: stacks/ltcd_85.cpp
//LeetCode - problem 85

class Solution { 
    int kadane(vector<int> &arr,int &up,int &down,int &n){
        int max_sum=INT_MIN,max_sum_end = 0;
        for(int i=0;i<n;i++){
            max_sum_end +=arr[i];
            if(max_sum_end<arr[i]){
                max_sum_end = arr[i];
                up=i;
            }
            if(max_sum<max_sum_end){
                max_sum = max_sum_end;
                down = i;
            }
        }
        return max_sum;
    }
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int cursum = 0, maxsum = 0,left=0,right=0,up=0,down=0;
        int row = matrix.size(),col = matrix[0].size();
        const int inf = -(row*col); 
        for(int l=0;l<col;l++){
            vector<int> temp(row);
            for(int r=l;r<col;r++){
                for(int i=0;i<row;i++){
                    temp[i]+=matrix[i][r]=='0'?inf:1;
                }
                cursum = kadane(temp,up,down,row);
                if(maxsum<cursum){
                    maxsum = cursum;
                    left = l;
                    right =r;
                }
            }
        }
        return maxsum;
    }
};
